src/simsopt/mhd/vmec_diagnostics.py

```diff
@@ -10,15 +10,277 @@
+    angles :math:`(\theta,\phi)`. In this case :math:`f` can be
+    written as a finite sum of squares:
+
+    .. math::
+        f = \sum_{s_j, \theta_j, \phi_j} R(s_j, \theta_k, \phi_{\ell})^2
+
+    where the :math:`\phi_{\ell}` grid covers a single field period.
+    Here, each residual term is
+
+    .. math::
+        R(\theta, \phi) = \sqrt{w_j \frac{n_{fp} \Delta_{\theta} \Delta_{\phi}}{V'}|\sqrt{g}|}
+        \frac{1}{B^3} \left( (N-\iota M)\vec{B}\times\nabla B\cdot\nabla\psi - (MG+NI)\vec{B}\cdot\nabla B \right).
+
+    Here, :math:`n_{fp}` is the number of field periods,
+    :math:`\Delta_{\theta}` and :math:`\Delta_{\phi}` are the spacing
+    of grid points in the poloidal and toroidal angles,
+    :math:`\sqrt{g} = 1/(\nabla s\cdot\nabla\theta \times
+    \nabla\phi)` is the Jacobian of the :math:`(s,\theta,\phi)`
+    coordinates, and :math:`V' = \int_0^{2\pi} d\theta \int_0^{2\pi}d\phi |\sqrt{g}| = dV/d\psi`
+    where :math:`V` is the volume enclosed by a flux surface.
+
+    Args:
+        vmec: A :obj:`simsopt.mhd.vmec.Vmec` object from which the
+          quasisymmetry error will be calculated.
+        s: Value of normalized toroidal flux at which you want the
+          quasisymmetry error evaluated, or a list of values. Each
+          value must be in the interval [0, 1], with 0 corresponding
+          to the magnetic axis and 1 to the VMEC plasma boundary.
+          This parameter corresponds to :math:`s_j` above.
+        m: Desired poloidal mode number :math:`M` in the magnetic field
+          strength :math:`B`, so
+          :math:`B = B(s, M \vartheta - n_{fp} \hat{N} \varphi)`
+          where :math:`\vartheta` and :math:`\varphi` are Boozer angles.
+        n: Desired toroidal mode number :math:`\hat{N} = N / n_{fp}` in the magnetic field
+          strength :math:`B`, so
+          :math:`B = B(s, M \vartheta - n_{fp} \hat{N} \varphi)`
+          where :math:`\vartheta` and :math:`\varphi` are Boozer angles.
+          Note that the supplied value of ``n`` will be multiplied by
+          the number of field periods :math:`n_{fp}`, so typically
+          ``n`` should be +1 or -1 for quasi-helical symmetry.
+        weights: The list of weights :math:`w_j` for each flux surface.
+          If ``None``, a weight of :math:`w_j=1` will be used for
+          all surfaces.
+        ntheta: Number of grid points in :math:`\theta` used to
+          discretize the flux surface average.
+        nphi: Number of grid points per field period in :math:`\phi` used to
+          discretize the flux surface average.
+    """
+
+    def __init__(self,
+                 vmec: Vmec,
+                 s: Union[float, RealArray],
+                 m: int = 1,
+                 n: int = 0,
+                 weights: RealArray = None,
+                 ntheta: int = 63,
+                 nphi: int = 64) -> None:
+
+        self.vmec = vmec
+        self.depends_on = ["vmec"]
```

Comment:
Line 116 implies that this class needs to be modified for graph framework. I'll take care of it after merging into the master.